rfcs/20210129-optimize-pd-creation/README.md

```diff
@@ -0,0 +1,262 @@
+
+#### Forward Primitive Descriptor Hint
+
+For some primitives the hint can affect equality of primitive descriptors.
+Currently, there are only two such primitives: pooling and shuffle. Depending on the
+hint the primitive descriptors can have different `diff_src` and `diff_dst`.
+To differentiate such primitives in the primitive cache the key contains
+corresponding memory descriptors. For pooling and shuffle the key contains a
+vector which contains `diff_src` and `diff_dst` which are obtained
+from the primitive descriptor.
+
+When creating a key to find primitive descriptor in the cache for primitives the
+hint can be used to deduce `diff_src` and `diff_dst` memory descriptors.
+Another option would be cloning the hint so that it can be used for creating key
+in the cache for primitives. This introduces overhead but given that it's almost
+completely reduced  with the optimization described in this RFC the options is
+reasonable. Though, there are only two use cases (primitives) and at this point
+the better option would be to limit impact to changing logic for those two primitives.
+
+#### Operation Descriptor for Sum, Concat and Reorder
+
+The difference between these primitives and others is that they don't have operation
+descriptors. Since operation descriptor is part of the key all primitives have to
+have operation descriptors. To fulfil the requirement there were introduced
+internal operation descriptors to implement primitive cache.
+
+```cpp
+struct dnnl_reorder_desc_t {
+    dnnl_primitive_kind_t primitive_kind;
+    dnnl_memory_desc_t src_md;
+    dnnl_memory_desc_t dst_md;
+    dnnl_engine_kind_t src_engine_kind;
+    dnnl_engine_kind_t dst_engine_kind;
+};
+
+struct dnnl_concat_desc_t {
+    dnnl_primitive_kind_t primitive_kind;
+    dnnl_memory_desc_t dst_md;
+    dnnl_dim_t n;
+    dnnl_dim_t concat_dimension;
+    std::vector<dnnl_memory_desc_t> src_mds;
+};
+
+struct dnnl_sum_desc_t {
+    dnnl_primitive_kind_t primitive_kind;
+    dnnl_memory_desc_t dst_md;
+    dnnl_dim_t n;
+    std::vector<float> scales;
+    std::vector<dnnl_memory_desc_t> src_mds;
+};
+```
+
+These operation descriptors are created and filled in constructors of primitive descriptors
+for the corresponding primitives.
+
+The main problem with these operation descriptors is that they have to be created
+to create a key to find primitive descriptor in the cache of primitives. It brings
+unnecessary overhead.
+Since the primitive descriptors take ownership of all the data that is used to create
+the internal operation descriptors there is no need to copy the data. The idea
+is to replace deep copy with shallow copy for all data fields which are POD or
+class types. This would allow to create the operation descriptors for the key
+without undesired overhead.
+
+The updated operation descriptors can look as follows:
+
+```cpp
+struct dnnl_reorder_desc_t {
+    dnnl_primitive_kind_t primitive_kind;
+    const dnnl_memory_desc_t *src_md;
```

Comment:
Does it mean that we will rely on `src_md` being alive until (at least) primitive creation?

Reply:
The `dnnl_reorder_desc_t ` will stored in `reorder_pd_t` which takes ownership of all the data. The lifetime of these entities are the same.

In case of creating a key to find primitive descriptor the `src_md` should be alive until we create a primitive descriptor which is the case.